File: app/static/js/meraki.js
import {api} from './api.js';
import {esc,attr,openModal,toast,navigate,confirmDialog} from './ui.js';

export async function merakiSettings(container){
  const clinics=await api.get('/api/clinics');
  const card=document.createElement('section');card.className='card mb';
  card.innerHTML=`<h2>Meraki · clinic-specific connections</h2><p>Each clinic has its own read-only API key. Choose a clinic to manage its key and map Meraki networks to its sites. No shared Meraki key is used.</p><label>Clinic<select data-clinic><option value="">Choose a clinic</option>${clinics.map(c=>`<option value="${c.id}">${esc(c.name)}</option>`).join('')}</select></label><button class="btn" data-open>Set up / import Meraki</button>`;
  (container.querySelector('section[aria-label="AI & integrations"]')||container).append(card);
  card.querySelector('[data-open]').onclick=()=>{const cid=Number(card.querySelector('[data-clinic]').value);if(cid)importMeraki(cid);else toast('Choose a clinic first.','error');};
}

export async function importMeraki(cid){
  const modal=openModal({title:'Meraki · clinic connection',size:'modal-lg',body:'Loading…',footer:'<button class="btn" data-close>Close</button>'});
  modal.root.querySelector('[data-close]').onclick=()=>modal.close();
  try{
    const [config,clinic,locations]=await Promise.all([api.get(`/api/meraki/clinics/${cid}/settings`),api.get(`/api/clinics/${cid}`),api.get(`/api/clinics/${cid}/locations`)]);
    modal.body.innerHTML=`<h3>${esc(clinic.name)}</h3><p>This key is used only for this clinic and its mapped sites. Use a dedicated Meraki Dashboard account with read-only access to the intended networks and enable Dashboard API access. AreaBook only sends GET requests.</p>
      <details ${config.configured?'':'open'}><summary>Clinic API key · ${config.configured?'saved':'not configured'}</summary><label>API key<input type="password" data-key autocomplete="off" placeholder="${config.configured?'Saved — leave blank to keep':'Enter clinic API key'}"></label><div class="actions"><button class="btn" data-save>Save clinic key</button><button class="btn" data-remove>Remove clinic key</button></div><p>The key is stored server-side. Protect the database and backups.</p></details>
      <button class="btn" data-test>Test key & load organizations</button><p role="status"></p><div data-select></div><div data-review></div>`;
    const q=s=>modal.body.querySelector(s),status=q('[role=status]');
    async function run(button,fn){button.disabled=true;try{await fn();}catch(e){status.textContent=e.message;}finally{button.disabled=false;}}
    function clear(){q('[data-select]').innerHTML='';q('[data-review]').innerHTML='';}
    q('[data-save]').onclick=e=>run(e.target,async()=>{const key=q('[data-key]').value.trim();if(!key){status.textContent='Enter a key to replace it; leave blank to keep the saved key.';return;}await api.put(`/api/meraki/clinics/${cid}/settings`,{api_key:key});q('[data-key]').value='';clear();status.textContent='Clinic key saved. Test it to choose an organization and network.';});
    q('[data-remove]').onclick=e=>run(e.target,async()=>{if(!await confirmDialog('Remove only this clinic’s Meraki key? Imported documentation stays intact; refreshes cannot run without a key.'))return;await api.put(`/api/meraki/clinics/${cid}/settings`,{api_key:''});q('[data-key]').value='';clear();status.textContent='Clinic key removed.';});
    q('[data-test]').onclick=e=>run(e.target,async()=>{
      clear();status.textContent='Reading organizations using this clinic’s saved key…';
      const orgs=await api.get(`/api/meraki/clinics/${cid}/organizations`);
      if(!orgs.length){status.textContent='No organizations visible to this key.';return;}
      status.textContent='Key accepted. Select the organization, network and matching local site.';
      q('[data-select]').innerHTML=`<label>Meraki organization<select data-org>${orgs.map(o=>`<option value="${attr(o.id)}">${esc(o.name)}</option>`).join('')}</select></label><label>Meraki network<select data-network></select></label><label>AreaBook site<select data-location><option value="">Main Site</option>${locations.map(l=>`<option value="${l.id}">${esc(l.name)}</option>`).join('')}</select></label><button class="btn btn-primary" data-preview>Build import preview</button>`;
      let networks=[];
      function mapping(){q('[data-review]').innerHTML='';const link=networks.find(n=>n.id===q('[data-network]').value)?.link;q('[data-location]').value=link?.location_id||'';q('[data-location]').disabled=Boolean(link);}
      async function load(){networks=[];q('[data-review]').innerHTML='';q('[data-preview]').disabled=true;q('[data-org]').disabled=true;q('[data-network]').innerHTML='';try{networks=await api.get(`/api/meraki/clinics/${cid}/networks`,{organization_id:q('[data-org]').value});q('[data-network]').innerHTML=networks.map(n=>`<option value="${attr(n.id)}">${esc(n.name)}${n.link?' · linked':''}</option>`).join('');mapping();status.textContent=networks.length?'Choose the correct site, then build a preview.':'No networks accessible in this organization.';}catch(e){status.textContent=e.message;}finally{q('[data-org]').disabled=false;q('[data-preview]').disabled=!networks.length;}}
      q('[data-org]').onchange=load;q('[data-network]').onchange=mapping;q('[data-location]').onchange=()=>q('[data-review]').innerHTML='';
      q('[data-preview]').onclick=e=>run(e.target,async()=>{
        const payload={clinic_id:cid,organization_id:q('[data-org]').value,network_id:q('[data-network]').value,location_id:Number(q('[data-location]').value)||null};
        q('[data-review]').innerHTML='';status.textContent='Reading inventory, ports and network observations. This can take up to four minutes…';
        const controls=[q('[data-org]'),q('[data-network]'),q('[data-location]'),q('[data-test]'),q('[data-save]'),q('[data-remove]')];controls.forEach(c=>c.disabled=true);
        try{const draft=await api.post('/api/meraki/preview',payload);review(draft,q('[data-review]'),status,modal);}finally{controls.forEach(c=>c.disabled=false);q('[data-location]').disabled=Boolean(networks.find(n=>n.id===payload.network_id)?.link);}
      });
      await load();
    });
  }catch(e){modal.body.textContent=e.message;}
}

function review(draft,container,status,modal){
  const machines=draft.records.filter(r=>r.proposal);
  status.textContent='Preview ready — nothing imported yet.';
  container.innerHTML=`<h3>Review ${esc(draft.site_name)}</h3><p>${machines.length} machines. Unique MAC matches reuse existing records; IP/name-only similarities need review. Existing device names, VM types, interfaces and wiring are preserved.</p><p>Clients were observed in the last 24 hours, not necessarily online now. Reported adapters are not assumed to be physical ports. Public IPv4 and APIPA are excluded; eligible IPv6 addresses are retained.</p>
    ${draft.warnings.map(w=>`<p class="text-muted">${esc(w)}</p>`).join('')}
    ${machines.map((r,i)=>`<details class="card mb"><summary>${esc(r.name)} · ${esc(r.proposal.action)} · ${esc(r.mac||'MAC unknown')}</summary><p>${esc(r.proposal.reason)}</p><p>IP: ${esc(r.addresses.map(a=>a.address).join(', ')||'No eligible address')} · ${esc(r.state||'State unknown')} · ${r.ports.length} ports</p><p>Reported attachment: ${esc(r.uplink||'Unknown')} · port ${esc(r.uplink_port||'unknown')} · VLAN ${esc(r.vlan||'unknown')}</p><label>Decision<select data-choice="${i}"><option value="skip" ${r.proposal.action==='skip'?'selected':''}>Skip local changes</option><option value="create" ${r.proposal.action==='create'?'selected':''}>Create device</option>${draft.choices.map(d=>`<option value="${d.id}" ${r.proposal.action==='match'&&r.proposal.device_id===d.id?'selected':''}>Match ${esc(d.name)} (#${d.id})</option>`).join('')}</select></label><details><summary>All collected details</summary><pre style="white-space:pre-wrap">${esc(JSON.stringify(r,null,2))}</pre></details></details>`).join('')}
    <details><summary>VLAN, VPN and LLDP/CDP observations</summary><pre style="white-space:pre-wrap">${esc(JSON.stringify(draft.records.filter(r=>!r.proposal),null,2))}</pre></details>
    <label><input type="checkbox" data-vlans checked> Add missing VLANs/subnets. Existing VLAN definitions and memberships stay unchanged.</label><label><input type="checkbox" data-uplinks> Fill missing uplinks from online clients with explicit attachment reports. VM hosts and existing wiring stay unchanged; exact port assignments remain observations.</label>
    <p>Switch ports are added on new switches only. Port configuration, VPN and undirected LLDP/CDP links remain source observations for review, not automatic local configuration.</p><button class="btn btn-primary" data-confirm>Confirm import into AreaBook</button>`;
  container.querySelector('[data-confirm]').onclick=async e=>{
    e.target.disabled=true;
    const decisions=machines.map((r,i)=>{const v=container.querySelector(`[data-choice="${i}"]`).value;return {kind:r.kind,id:r.id,action:['skip','create'].includes(v)?v:'match',device_id:Number(v)||null};});
    try{const result=await api.post('/api/meraki/import',{token:draft.token,decisions,import_networks:container.querySelector('[data-vlans]').checked,import_uplinks:container.querySelector('[data-uplinks]').checked});modal.close();toast(`Meraki: ${result.created} created, ${result.matched} matched, ${result.vlans_added} VLANs added.`,'success');navigate('#/clinics/'+result.clinic_id);}catch(error){status.textContent=error.message;e.target.disabled=false;}
  };
}

export async function merakiClinic(container,cid){
  const user=await api.get('/api/auth/me');const admin=user.roles.includes('admin');if(!admin&&user.active_role!=='it')return;
  const data=await api.get(`/api/meraki/clinics/${cid}`);if(!container.isConnected||(!admin&&!data.sites.length))return;
  const card=document.createElement('section');card.className='card mb';
  card.innerHTML=`<h2>Meraki network observations</h2><p>Clinic-specific read-only connection. Snapshots do not prove current reachability; missing devices are never deleted automatically.</p>${admin?'<button class="btn" data-setup>Set up / import Meraki</button>':''}${data.sites.map(s=>`<details><summary>${esc(s.name)} · ${esc(s.updated_at)} UTC</summary>${JSON.parse(s.warnings).map(w=>`<p>${esc(w)}</p>`).join('')}${data.records.filter(r=>r.network_id===s.network_id).map(r=>`<details><summary>${esc(r.data.name)} · ${esc(r.data.kind)}${r.local_id?' · linked #'+r.local_id:''}</summary><p>Last observation: ${esc(r.updated_at)} UTC</p><pre style="white-space:pre-wrap">${esc(JSON.stringify(r.data,null,2))}</pre></details>`).join('')}</details>`).join('')}`;
  container.append(card);if(admin)card.querySelector('[data-setup]').onclick=()=>importMeraki(cid);
}

File: app/routers/meraki.py
"""Clinic-specific credentials and explicit reviewed Meraki network imports."""
import json
import secrets
import time
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field
from ..database import db_dependency
from ..meraki_client import Client, ID
from ..meraki_mapping import collect, identity
from ..unifi_mapping import local_state, match, text
from .unifi import target, fill_device, Import

router=APIRouter(prefix='/api/meraki',tags=['Meraki'])
SCHEMA='''
CREATE TABLE IF NOT EXISTS meraki_credentials (
 id INTEGER PRIMARY KEY, clinic_id INTEGER NOT NULL UNIQUE REFERENCES clinics(id) ON DELETE CASCADE,
 api_key TEXT NOT NULL, generation TEXT NOT NULL);
CREATE TABLE IF NOT EXISTS meraki_sites (
 id INTEGER PRIMARY KEY, organization_id TEXT NOT NULL, network_id TEXT NOT NULL UNIQUE,
 clinic_id INTEGER NOT NULL REFERENCES clinics(id) ON DELETE CASCADE,
 location_id INTEGER REFERENCES clinic_locations(id) ON DELETE CASCADE, name TEXT NOT NULL,
 warnings TEXT NOT NULL DEFAULT '[]', updated_at TEXT NOT NULL DEFAULT CURRENT_TIMESTAMP);
CREATE UNIQUE INDEX IF NOT EXISTS meraki_local_site ON meraki_sites(clinic_id,IFNULL(location_id,0));
CREATE TABLE IF NOT EXISTS meraki_records (
 id INTEGER PRIMARY KEY, network_id TEXT NOT NULL, kind TEXT NOT NULL, external_id TEXT NOT NULL,
 clinic_id INTEGER NOT NULL REFERENCES clinics(id) ON DELETE CASCADE,
 local_id INTEGER, data TEXT NOT NULL, updated_at TEXT NOT NULL DEFAULT CURRENT_TIMESTAMP,
 UNIQUE(network_id,kind,external_id));
CREATE TABLE IF NOT EXISTS meraki_previews (
 token TEXT PRIMARY KEY, clinic_id INTEGER NOT NULL REFERENCES clinics(id) ON DELETE CASCADE,
 actor_id INTEGER NOT NULL, expires REAL NOT NULL, data TEXT NOT NULL);
'''

def admin(c):
    if 'admin' not in c.user['roles']:raise HTTPException(403,'Meraki setup and imports require an administrator')

def credential(c,cid):
    row=c.execute('SELECT api_key,generation FROM meraki_credentials WHERE clinic_id=?',(cid,)).fetchone()
    if not row or not row['api_key']:raise HTTPException(409,'Save a Meraki API key for this clinic first')
    return row

@router.get('/clinics/{cid}/settings')
def settings(cid:int,c=Depends(db_dependency)):
    admin(c);target(c,cid,None)
    row=c.execute('SELECT api_key FROM meraki_credentials WHERE clinic_id=?',(cid,)).fetchone()
    return {'configured':bool(row and row['api_key'])}

class Config(BaseModel):api_key:str=Field(max_length=2000)

@router.put('/clinics/{cid}/settings')
def configure(cid:int,p:Config,c=Depends(db_dependency)):
    admin(c);target(c,cid,None);key=p.api_key.strip()
    if any(ord(ch)<33 or ord(ch)>126 for ch in key):raise HTTPException(422,'API key must not contain spaces or control characters')
    c.execute('INSERT INTO meraki_credentials(clinic_id,api_key,generation) VALUES (?,?,?) ON CONFLICT(clinic_id) DO UPDATE SET api_key=excluded.api_key,generation=excluded.generation',(cid,key,secrets.token_hex(16)))
    c.execute('DELETE FROM meraki_previews WHERE clinic_id=?',(cid,))
    return settings(cid,c)

@router.get('/clinics/{cid}/organizations')
def organizations(cid:int,c=Depends(db_dependency)):
    admin(c);target(c,cid,None)
    return [{'id':identity(r.get('id')),'name':text(r.get('name'))} for r in Client(credential(c,cid)['api_key']).collection('/organizations')]

@router.get('/clinics/{cid}/networks')
def networks(cid:int,organization_id:str,c=Depends(db_dependency)):
    admin(c);target(c,cid,None);identity(organization_id)
    rows=Client(credential(c,cid)['api_key']).collection('/organizations/'+organization_id+'/networks',{'perPage':1000})
    links={r['network_id']:dict(r) for r in c.execute('SELECT * FROM meraki_sites WHERE clinic_id=?',(cid,))}
    return [{'id':identity(r.get('id')),'name':text(r.get('name')),'link':links.get(str(r.get('id')))} for r in rows if str(r.get('organizationId'))==organization_id]

def check_link(c,org,network,cid,location):
    if c.execute('SELECT id FROM meraki_records WHERE network_id=? AND clinic_id<>? LIMIT 1',(network,cid)).fetchone():raise HTTPException(409,'This network already belongs to another clinic')
    for r in c.execute('SELECT * FROM meraki_sites WHERE network_id=? OR (clinic_id=? AND location_id IS ?)',(network,cid,location)):
        if (r['organization_id'],r['network_id'],r['clinic_id'],r['location_id'])!=(org,network,cid,location):
            raise HTTPException(409,'This Meraki network or AreaBook site is already mapped. Use the existing mapping.')

class Preview(BaseModel):
    clinic_id:int=Field(gt=0)
    organization_id:str=Field(pattern='^'+ID+'$')
    network_id:str=Field(pattern='^'+ID+'$')
    location_id:int|None=Field(default=None,gt=0)

@router.post('/preview')
def preview(p:Preview,c=Depends(db_dependency)):
    admin(c);target(c,p.clinic_id,p.location_id);check_link(c,p.organization_id,p.network_id,p.clinic_id,p.location_id)
    key=credential(c,p.clinic_id)
    data=collect(Client(key['api_key']),p.organization_id,p.network_id)
    local,fingerprint=local_state(c,p.clinic_id,p.location_id)
    for r in data['records']:
        if r['kind'] not in ('devices','clients'):continue
        old=c.execute('SELECT local_id FROM meraki_records WHERE network_id=? AND kind=? AND external_id=?',(p.network_id,r['kind'],r['id'])).fetchone()
        r['proposal']=match(r,local,old['local_id'] if old else None)
        r['proposal']['reason']=r['proposal']['reason'].replace('UniFi','Meraki')
    saved=p.model_dump()|data|{'generation':key['generation'],'fingerprint':fingerprint,'choices':[{'id':d['id'],'name':d['name']} for d in local]}
    token=secrets.token_urlsafe(32)
    c.execute('DELETE FROM meraki_previews WHERE expires<?',(time.time(),))
    c.execute('INSERT INTO meraki_previews VALUES (?,?,?,?,?)',(token,p.clinic_id,c.user['id'],time.time()+1800,json.dumps(saved)))
    return {k:v for k,v in saved.items() if k not in ('generation','fingerprint')}|{'token':token}

@router.post('/import')
def import_site(p:Import,c=Depends(db_dependency)):
    admin(c)
    saved=c.execute('SELECT * FROM meraki_previews WHERE token=? AND actor_id=? AND expires>?',(p.token,c.user['id'],time.time())).fetchone()
    if not saved:raise HTTPException(409,'Preview expired or already used; build a new preview')
    data=json.loads(saved['data']);cid=data['clinic_id'];location=data['location_id'];network=data['network_id']
    if credential(c,cid)['generation']!=data['generation']:raise HTTPException(409,'Clinic credentials changed; build a new preview')
    target(c,cid,location);check_link(c,data['organization_id'],network,cid,location)
    local,fingerprint=local_state(c,cid,location)
    if fingerprint!=data['fingerprint']:raise HTTPException(409,'Local topology changed; build a new preview')
    records={(r['kind'],r['id']):r for r in data['records'] if r['kind'] in ('devices','clients')}
    choices={(d.kind,d.id):d for d in p.decisions}
    if len(choices)!=len(p.decisions) or set(choices)!=set(records):raise HTTPException(422,'Review every device exactly once')
    selected={};counts={'created':0,'matched':0,'skipped':0,'vlans_added':0,'uplinks_added':0}
    for k,r in records.items():
        choice=choices[k]
        if choice.action=='skip':counts['skipped']+=1;continue
        old=c.execute('SELECT local_id FROM meraki_records WHERE network_id=? AND kind=? AND external_id=?',(network,*k)).fetchone()
        if old and old['local_id'] is not None and (choice.action!='match' or choice.device_id!=old['local_id']):raise HTTPException(409,'Existing source link cannot be reassigned or duplicated')
        if choice.action=='match':
            if not any(d['id']==choice.device_id for d in local):raise HTTPException(422,'Match must belong to this clinic/site')
            did=choice.device_id;counts['matched']+=1
        else:
            if r['proposal']['action']=='match':raise HTTPException(409,'Exact match exists; match or skip')
            did=c.execute('INSERT INTO devices(clinic_id,location_id,name,device_type,manufacturer,os,serial,notes) VALUES (?,?,?,?,?,?,?,?)',
                (cid,location,r['name'],r['device_type'],r['manufacturer'],r['os'],r['serial'],'Imported from Meraki. Review device type and network observations.')).lastrowid
            counts['created']+=1
        fill_device(c,did,r|{'ports':[]},choice.action=='create','Meraki')
        if choice.action=='create':
            for port in r['ports']:
                c.execute('INSERT INTO network_interfaces(device_id,name,port_group,port_order,notes) VALUES (?,?,?,?,?)',
                    (did,'Port '+port['id']+(' · '+port['name'] if port['name'] else ''),'Meraki ports',port['idx'],
                     'Reported by Meraki. Port configuration snapshot: '+json.dumps(port)+'. Review VLAN membership; negotiation is not a supported-speed inventory.'))
        selected[k]=did
    if p.import_uplinks:
        # Only explicit client attachment reports; never orient undirected LLDP links.
        for k,did in selected.items():
            r=records[k];up=selected.get(('devices',r['uplink']))
            d=c.execute('SELECT * FROM devices WHERE id=?',(did,)).fetchone()
            if k[0]!='clients' or r['state']!='Online' or r['connection_type'] not in ('WIRED','WIRELESS') or not up or up==did or d['uplink_id'] or d['device_type']=='vm':continue
            if c.execute('SELECT id FROM device_links WHERE device_id=?',(did,)).fetchone():continue
            pending=[up];seen=set();cycle=False
            while pending:
                n=pending.pop()
                if n==did:cycle=True;break
                if n in seen:continue
                seen.add(n)
                parent=c.execute('SELECT uplink_id FROM devices WHERE id=?',(n,)).fetchone()
                if parent and parent['uplink_id']:pending.append(parent['uplink_id'])
                pending.extend(row['uplink_id'] for row in c.execute('SELECT uplink_id FROM device_links WHERE device_id=?',(n,)))
            if cycle:continue
            c.execute('UPDATE devices SET uplink_id=?,link_type=? WHERE id=?',(up,'wireless' if r['connection_type']=='WIRELESS' else 'ethernet',did));counts['uplinks_added']+=1
    for r in data['records']:
        local_id=selected.get((r['kind'],r['id']))
        if r['kind']=='networks' and p.import_networks and r['tag']:
            existing=c.execute('SELECT id FROM vlans WHERE clinic_id=? AND location_id IS ? AND tag=?',(cid,location,r['tag'])).fetchone()
            if existing:local_id=existing['id']
            else:
                local_id=c.execute('INSERT INTO vlans(clinic_id,location_id,tag,name,subnets,notes) VALUES (?,?,?,?,?,?)',(cid,location,r['tag'],r['name'] or 'Meraki VLAN',json.dumps(r['subnets']),'Imported from Meraki. Review controller and interface membership.')).lastrowid
                counts['vlans_added']+=1
        c.execute('''INSERT INTO meraki_records(network_id,kind,external_id,clinic_id,local_id,data) VALUES (?,?,?,?,?,?)
            ON CONFLICT(network_id,kind,external_id) DO UPDATE SET data=excluded.data,local_id=COALESCE(meraki_records.local_id,excluded.local_id),updated_at=CURRENT_TIMESTAMP''',(network,r['kind'],r['id'],cid,local_id,json.dumps(r)))
    c.execute('''INSERT INTO meraki_sites(organization_id,network_id,clinic_id,location_id,name,warnings) VALUES (?,?,?,?,?,?)
        ON CONFLICT(network_id) DO UPDATE SET name=excluded.name,warnings=excluded.warnings,updated_at=CURRENT_TIMESTAMP''',(data['organization_id'],network,cid,location,data['site_name'],json.dumps(data['warnings'])))
    c.execute('DELETE FROM meraki_previews WHERE token=?',(p.token,))
    from ..integration_sync import discover
    discover(c)
    return counts|{'clinic_id':cid,'warnings':data['warnings']}

@router.get('/clinics/{cid}')
def observations(cid:int,c=Depends(db_dependency)):
    target(c,cid,None)
    return {'sites':[dict(r) for r in c.execute('SELECT * FROM meraki_sites WHERE clinic_id=?',(cid,))],
            'records':[{'data':json.loads(r['data']),'local_id':r['local_id'],'updated_at':r['updated_at'],'network_id':r['network_id']} for r in c.execute('SELECT * FROM meraki_records WHERE clinic_id=?',(cid,))]}
